backend-services/configuration/core/configuration.py

```diff
@@ -0,0 +1,1531 @@
+        app = self.make_rest_app()
+        app.listen(REST_PORT)
+        log.info("REST worker started and listening to port {}".format(REST_PORT))
+        IOLoop.current().start()
+
+
+class ConfigurationDataWorker(ConsumerProducerMixin):
+    """
+    RabbitMQ Consumer/Producer for the Configuration Service.
+    """
+
+    def __init__(
+        self, connection: Connection, shared_memory_manager_dict: Dict
+    ) -> NoReturn:
+        self.connection = connection
+        self.shared_memory_manager_dict = shared_memory_manager_dict
+        self.redis = redis.Redis(host=REDIS_HOST, port=REDIS_PORT)
+        ping_redis(self.redis)
+
+        # EXCHANGES
+        self.autoconf_exchange = create_exchange("autoconf", connection, declare=True)
+        self.command_exchange = create_exchange("command", connection, declare=True)
+
+        # QUEUES
+        self.autoconf_filtered_update_queue = create_queue(
+            SERVICE_NAME,
+            exchange=self.autoconf_exchange,
+            routing_key="filtered-update",
+            priority=4,
+            random=True,
+        )
+        self.stop_queue = create_queue(
+            "{}-{}".format(SERVICE_NAME, uuid()),
+            exchange=self.command_exchange,
+            routing_key="stop-{}".format(SERVICE_NAME),
+            priority=1,
+        )
+
+        log.info("data worker initiated")
+
+    def get_consumers(self, Consumer: Consumer, channel: Connection) -> List[Consumer]:
+        return [
+            Consumer(
+                queues=[self.autoconf_filtered_update_queue],
+                on_message=self.handle_filtered_autoconf_updates,
+                prefetch_count=1,
+                accept=["ujson"],
+            ),
+            Consumer(
+                queues=[self.stop_queue],
+                on_message=self.stop_consumer_loop,
+                prefetch_count=100,
+                accept=["ujson"],
+            ),
+        ]
+
+    def handle_filtered_autoconf_updates(self, message):
+        """
+        Receives a "autoconf-update" message batch (filtered by the prefixtree),
+        translates the corresponding BGP updates into ARTEMIS configuration
+        and rewrites the configuration
+        :param message:
+        :return:
+        """
+        if not message.acknowledged:
+            message.ack()
+        try:
+            bgp_updates = message.payload
+            if not isinstance(bgp_updates, list):
+                bgp_updates = [bgp_updates]
+
+            # load initial YAML configuration from file
+            with open(self.shared_memory_manager_dict["config_file"], "r") as f:
+                raw = f.read()
+                yaml_conf = ruamel.yaml.load(
+                    raw, Loader=ruamel.yaml.RoundTripLoader, preserve_quotes=True
+                )
+
+            # save initial file content to ensure that it has not changed while processing
+            with open(self.shared_memory_manager_dict["config_file"], "r") as f:
```

Comment:
why do we reopen the file from above? can't we move this line above?

Reply:
In the first instance we do an `f.read()` to get the raw content and  turn it into yaml.
In the second instance we read the lines (`f.readlines()`) so that we can do an easy diff check.
So I think both are needed because the file pointer progresses with the first call.

Reply:
so in that case will something like `f.seek(0,0)` to return back to the start of the file will be ok?

Reply:
Comment is now outdated.